### PR TITLE
Fix incorrect perplexity reported when label smoothing enabled.

### PR DESCRIPTION
**Pull request by helson73**
Two updates of label smoothing:
1. Report true perplexity.
2. Reduce computation cost of mask operation.

### CODE REVIEW

**Comment by helson73** [comment]
Before:
Train perplexity: 32.4537
Train accuracy: 38.7784
**Validation perplexity: 56.2081**
**Validation accuracy: 33.0021**

After PR #456:
Train perplexity: 29.8781
Train accuracy: 40.5064
**Validation perplexity: 52.0337**
**Validation accuracy: 34.0583**

After PR #462:
Train perplexity: 79.9049 - exp(xentropy)
Train accuracy: 40.6099
Validation perplexity: 128.389 - exp(xentropy)
**Validation accuracy: 34.5876**

This update:
Train perplexity: 32.3073
Train accuracy: 40.675
**Validation perplexity: 51.5402**
**Validation accuracy: 34.6887**

Looks much better now. Training PPL became worse but validation PPL is better. Quite good regularization !
Next I will implement token-based batch. 
To repeat exact experiment "Attention is All You Need" did, following updates are needed.
1) Batch size should be defined as number of tokens, rather than number of sentences. Therefore, number of sentences in one batch would be varied.
2) Learning schedule (including updating learning rate) should be managed on update iterations, rather than epochs.

**Comment by JianyuZhan** [comment]
Nice, thank you @helson73 !

**Comment by vince62s** [comment]
If you really want to replicate Attention is All You need, you will really need multi GPU sync working.
results are so hyper-dependent on the batchsize / number of GPU, we will need this multi GPU working.

**Comment by helson73** [comment]
@vince62s Yes, I need more time to do that since I am new to pytorch. 
Even though multi-GPU is supported, find or prepare a device that has 8 GPUs is not so easy. They reported each batch contains 25k tokens, approximately 500~1000 sentences, not usual in other work. 
But I think if we get close result with T2T they released under same condition, our goal is somewhat achieved.

**Comment by vince62s** [comment]
I recently tested fairseq-py and it's really fast and multi-gpu
Maybe you could take some stuff from their repo on this multi gpu stuff.

**Comment by srush** [comment]
I think we can make progress without multi-gpu. We'll get to that, but the changes you are making a really useful with multi-gpu.
